### ISSUE TITLE
Installed packages with custom cflags/ldflags fail on second env install

### ISSUE THREAD

**Issue by rbberger** [LLNL/spack#35913]
### Steps to reproduce

```console
$ # works on first attempt (using some nonsense example for illustration)
$ spack env create testing
$ spack env activate testing
$ spack add zlib ldflags="-lm"
$ spack install

$ # fails in all future installs
$ spack env create testing2
$ spack env activate testing2
$ spack add zlib ldflags="-lm"
$ spack install
```

### Error message

```console
$ spack --debug --stacktrace install
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.193755] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.225072] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.232216] Reading config from file /vast/home/rberger/projects/devops/spack/var/spack/environments/second-testing/spack.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.235824] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/environment/environment.py:201 ==> [2023-03-07-09:38:59.260875] Using environment 'second-testing'
lib/spack/spack/cmd/__init__.py:125 ==> [2023-03-07-09:38:59.286021] Imported install from built-in commands
lib/spack/spack/cmd/__init__.py:125 ==> [2023-03-07-09:38:59.287783] Imported install from built-in commands
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.294027] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.307555] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.316175] Reading config from file /vast/home/rberger/projects/devops/spack/var/spack/environments/second-testing/spack.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.319884] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/environment/environment.py:225 ==> [2023-03-07-09:38:59.344726] Deactivated environment 'second-testing'
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.361988] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.364852] Reading config from file /vast/home/rberger/.spack/repos.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.370922] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/database.py:391 ==> [2023-03-07-09:38:59.396321] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/database.py:397 ==> [2023-03-07-09:38:59.396588] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/database.py:391 ==> [2023-03-07-09:38:59.398045] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/database.py:397 ==> [2023-03-07-09:38:59.398332] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/bootstrap/config.py:124 ==> [2023-03-07-09:38:59.398603] [BOOTSTRAP CONFIG SCOPE] name=_builtin
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.399339] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/config.py:135 ==> [2023-03-07-09:38:59.407274] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/vast/home/rberger/projects/devops/spack/etc/spack/defaults
lib/spack/spack/bootstrap/config.py:136 ==> [2023-03-07-09:38:59.407517] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/vast/home/rberger/projects/devops/spack/etc/spack/defaults/linux
lib/spack/spack/bootstrap/config.py:135 ==> [2023-03-07-09:38:59.407759] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/vast/home/rberger/.spack/bootstrap/config
lib/spack/spack/bootstrap/config.py:136 ==> [2023-03-07-09:38:59.407985] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/vast/home/rberger/.spack/bootstrap/config/linux
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.409632] Reading config from file /vast/home/rberger/.spack/bootstrap/config/linux/compilers.yaml
lib/spack/spack/bootstrap/_common.py:214 ==> [2023-03-07-09:38:59.459603] [BOOTSTRAP ROOT SPEC] patchelf@0.13.1: %gcc target=x86_64
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:38:59.460148] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/_common.py:176 ==> [2023-03-07-09:38:59.473585] [BOOTSTRAP EXECUTABLES patchelf] Try installed specs with query 'patchelf@0.13.1: %gcc target=x86_64'
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.128259] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.227237] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/modules.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.243787] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/linux/modules.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.378960] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/bootstrap/_common.py:214 ==> [2023-03-07-09:39:02.434116] [BOOTSTRAP ROOT SPEC] gnupg@2.3: %gcc target=x86_64
lib/spack/spack/bootstrap/_common.py:214 ==> [2023-03-07-09:39:02.450749] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
lib/spack/spack/bootstrap/core.py:357 ==> [2023-03-07-09:39:02.450978] [BOOTSTRAP MODULE clingo] Try importing from Python
lib/spack/spack/bootstrap/_common.py:73 ==> [2023-03-07-09:39:02.581374] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.6/prqkzynv2nwko5mktitebgkeumuxkveu" provides the "clingo" Python module
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.582737] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.612734] Reading config from file /vast/home/rberger/projects/devops/spack/var/spack/environments/second-testing/spack.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.616666] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/config.yaml
lib/spack/spack/environment/environment.py:201 ==> [2023-03-07-09:39:02.642265] Using environment 'second-testing'
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.642681] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.649020] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.651569] Reading config from file /vast/home/rberger/.spack/repos.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.757739] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:02.823786] Reading config from file /vast/home/rberger/projects/devops/spack/etc/spack/defaults/mirrors.yaml
lib/spack/spack/util/web.py:330 ==> [2023-03-07-09:39:02.826435] Checking existence of https://mirror.spack.io/build_cache/index.json
lib/spack/spack/util/web.py:358 ==> [2023-03-07-09:39:03.005559] Failure reading URL: HTTP Error 404: Not Found
lib/spack/spack/config.py:1003 ==> [2023-03-07-09:39:03.016061] Reading config from file /vast/home/rberger/.spack/linux/compilers.yaml
Traceback (most recent call last):
  File "/vast/home/rberger/projects/devops/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack_installable/main.py", line 37, in main
    sys.exit(spack.main.main(argv))
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/main.py", line 1004, in main
    return _main(argv)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/main.py", line 959, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/main.py", line 987, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/main.py", line 636, in _invoke_command
    return_val = command(parser, args)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/cmd/install.py", line 511, in install
    reporter_factory=reporter_factory,
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/cmd/install.py", line 380, in install_all_specs_from_active_environment
    concretized_specs = env.concretize(tests=install_kwargs["tests"])
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/environment/environment.py", line 1325, in concretize
    return self._concretize_together(tests=tests)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/environment/environment.py", line 1378, in _concretize_together
    *self.user_specs, tests=tests
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/concretize.py", line 739, in concretize_specs_together
    return _concretize_specs_together_new(*abstract_specs, **kwargs)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/concretize.py", line 746, in _concretize_specs_together_new
    result = solver.solve(abstract_specs, tests=kwargs.get("tests", False))
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/solver/asp.py", line 2539, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/solver/asp.py", line 764, in solve
    answers = builder.build_specs(spec_attrs)
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/solver/asp.py", line 2411, in build_specs
    self.reorder_flags()
  File "/vast/home/rberger/projects/devops/spack/lib/spack/spack/solver/asp.py", line 2328, in reorder_flags
    assert set(compiler_flags) == set(ordered_compiler_flags), msg
AssertionError: {'-lm'} does not equal set()
```

### Information on your system

* **Spack:** 0.20.0.dev0 (6f396aff99a3747e9d0a36c26ad788502648c52b)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-haswell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by rbberger** [comment]
Also happens with `cflags`, e.g. `cflags="-fPIC"`

**Comment by alalazo** [comment]
Fixed in #35951
